Banner validation (Backend)

eulerbutcooler opened the pull request:
### Banner backend validations.

This PR addresses the issue **#388** 

I've used express validator for the backend validation checks.

Added Changes: 

- Added all the validation logic for banners in **banner.
- Reused validation logic for **/add_banner** and **/edit_banner** 
- Reused id check from param in both **/delete_banner** and **/get_banner/:id**
- Added url check for **/get_banner_by_url**
- Added checks for url that are triggered only when the close action button is set to 'open url' or 'open url in a new tab'
- Consistent background and font color checks allowing only hex color codes
- Added/Updated error messages expected in **/e2e/banner.test.mjs** and **unit/controllers/banner.test.js**

Removed code:

- Validation checks from **banner.controller.js**
- Validation checks from **Banner.js**

I've left the validatePosition, validatePositionWrapper, etc as they might still be used by some other service.

eulerbutcooler commented on backend/src/routes/banner.routes.js:
```diff
@@ -1,18 +1,47 @@
+  handleValidationErrors,
+  bannerController.getBannerById
+);
+router.get('/get_banner_by_url', bannerController.getBannerByUrl);
```
Added these changes in the latest commit

eulerbutcooler commented on frontend/src/services/bannerServices.js:
```diff
@@ -1,51 +1,65 @@
+    const response = await apiClient.post('/banner/add_banner', bannerData);
+    return response.data;
+  } catch (error) {
+    console.error('Add Banner error:', error.response.data.errors);
```
It'll always exist in that form

eulerbutcooler commented on backend/src/utils/banner.helper.js:
```diff
@@ -1,11 +1,65 @@
+    .bail()
+    .isIn(['no action', 'open url', 'open url in a new tab'])
+    .withMessage('Invalid close button action'),
+  body('url')
+    .optional()
+    .custom((value, { req }) => {
+      if (!value && ['open url', 'open url in a new tab'].includes(req.body.closeButtonAction)) {
+        return false;
+      }
+      return true;
+    })
+    .withMessage('URL is required when close button action is set to open URL')
+    .bail()
+    .custom((value) => {
+      if (!value) return true;
+      try {
+        const url = new URL(value);
+        return ['http:', 'https:'].includes(url.protocol);
+      } catch {
+        return false;
+      }
+    })
+    .withMessage('URL must use HTTP or HTTPS protocol'),
```
Have left those functions as they might be used elsewhere

eulerbutcooler commented on backend/src/controllers/banner.controller.js:
```diff
@@ -1,246 +1,95 @@
+      const deletionResult = await bannerService.deleteBanner(id);
+
+      if (!deletionResult) {
+        return res.status(400).json({
+          errors: [{ msg: 'Banner with the specified id does not exist' }],
+        });
+      }
```
Not needed

eulerbutcooler commented:
@erenfn Let me know if i need to make any changes here

erenfn reviewed (changes requested):
Missing repetition type returns:

```
{
    "errors": [
        "Invalid value",
        "Invalid repetition"
    ]
}
```

However, missing position returns

```
{
    "errors": [
        "Position is required"
    ]
}
```

In repetitionType, can we remove "Invalid value", and change Invalid repetition to "Repetition type is required"?

eulerbutcooler commented:
@erenfn fixed the error message as requested

erenfn commented:
> @erenfn fixed the error message as requested i noticed there are no tests for repetition type. Do you want me to add those?

Sure. That would be great. Also can you change the validation of repetitionType to this?

```
  body('repetitionType')
    .custom((value) => {
      if (!value) {
        throw new Error('Repetition type is required');
      }
      if (!['show only once', 'show every visit'].includes(value)) {
        throw new Error('Invalid repetition type');
      }
      return true;
    }),
```

eulerbutcooler commented:
> > @erenfn fixed the error message as requested i noticed there are no tests for repetition type. Do you want me to add those?
> 
> Sure. That would be great. Also can you change the validation of repetitionType to this?
> 
> ```
>   body('repetitionType')
>     .custom((value) => {
>       if (!value) {
>         throw new Error('Repetition type is required');
>       }
>       if (!['show only once', 'show every visit'].includes(value)) {
>         throw new Error('Invalid repetition type');
>       }
>       return true;
>     }),
> ```

Added the tests for both add_banner and edit_banner and also implemented the repetitionType validation as requested

erenfn reviewed (approved):
Looks good!
